ethereum/eth/src/main/java/tech/pegasys/pantheon/ethereum/eth/manager/bounded/BoundedTimedQueue.java

```diff
@@ -0,0 +1,163 @@
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.eth.manager.bounded;
+
+import static java.util.Objects.requireNonNull;
+
+import tech.pegasys.pantheon.metrics.Counter;
+import tech.pegasys.pantheon.metrics.MetricsSystem;
+import tech.pegasys.pantheon.metrics.PantheonMetricCategory;
+import tech.pegasys.pantheon.metrics.noop.NoOpMetricsSystem;
+
+import java.time.temporal.ChronoUnit;
+import java.time.temporal.TemporalUnit;
+import java.util.Locale;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class BoundedTimedQueue extends LinkedBlockingDeque<Runnable> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final long keepAlive;
+  private final TemporalUnit keepAliveUnit;
+  private final long evictionPollTimeout;
+  private final TimeUnit evictionPollTimeoutUnit;
+  private final MetricsSystem metricsSystem;
+  private final String metricName;
+  private final Counter totalEvictedTaskCounter;
+
+  public BoundedTimedQueue() {
+    this(
+        Default.CAPACITY,
+        Default.KEEP_ALIVE,
+        Default.KEEP_ALIVE_UNIT,
+        Default.EVICTION_POLL_TIMEOUT,
+        Default.EVICTION_POLL_TIMEOUT_UNIT,
+        Default.METRICS_SYSTEM);
+  }
+
+  public BoundedTimedQueue(
+      final int capacity,
+      final long keepAlive,
+      final TemporalUnit keepAliveUnit,
+      final long evictionPollTimeout,
+      final TimeUnit evictionPollTimeoutUnit,
+      final MetricsSystem metricsSystem) {
+    super(capacity);
+    this.keepAlive = keepAlive;
+    this.keepAliveUnit = keepAliveUnit;
+    this.evictionPollTimeout = evictionPollTimeout;
+    this.evictionPollTimeoutUnit = evictionPollTimeoutUnit;
+    this.metricsSystem = metricsSystem;
+    this.metricName = this.getClass().getSimpleName().toLowerCase(Locale.US).replace('-', '_');
+    this.totalEvictedTaskCounter =
+        this.metricsSystem.createCounter(
+            PantheonMetricCategory.EXECUTORS,
+            metricName + "_rejected_tasks_total",
+            "Total number of tasks rejected by this working queue.");
+  }
+
+  @Override
+  public boolean offer(final Runnable runnable) {
+    try {
+      requireNonNull(runnable);
+      if (!(runnable instanceof TimedTask)) {
+        return false;
+      }
+      final TimedTask task = (TimedTask) runnable;
+      if (!isEligible(task)) {
+        LOG.debug("Task does not match time policy, rejecting task.");
+        return false;
+      }
+      final boolean offered = super.offer(task);
+      if (offered) {
+        return true;
+      }
+      // Try to evict tasks based on time policy.
+      int evictedTaskCounter = processEviction();
+      if (evictedTaskCounter == 0) {
+        // No task has been evicted, rejecting new task.
+        LOG.warn(
```

Comment:
We need some kind of rate limiting for log messages when messages are dropped - otherwise we'll try to print hundreds of thousands of log messages which will quickly become the performance bottleneck and make things worse.

I'm not entirely sure what the right policy is for this is - we do want to make it visible that something is going wrong, but definitely need to avoid logging for every dropped message.